File: docs/palindromes-in-peg/chain_finite.py
"""Finite right-dp / three-way chain monitor, composed after local DP search.

The caller supplies right-side places C+1,C+2,... through a one-cell PORT,
only at a ready state. Setup is offline. Each subsequent place takes
constant local work: PERIOD bounces along a marked semiperiod, while
WINDOW moves left for ordinary palindrome matching. This implements the
branch decisions, not the outer center shift/restart/move scheduler.
"""
import copy

from fpp_finite import Program, LEFT, END, BLANK
from dp_search_finite import build_search_program, WINDOW, STATUS, center_symbol, period_symbol

PERIOD, PORT = 16, 17


def first(symbol):
  return "F:" + symbol


def last(symbol):
  return "T:" + symbol


def build_chain_program(alphabet="abs"):
  search = build_search_program(alphabet)
  p = Program(search.alphabet, ntapes=18)
  p.source_alphabet = search.source_alphabet
  p.scalar_tapes = (STATUS, PORT)
  p.code = copy.deepcopy(search.code)
  p.outcomes = {search.missed: "no_chain"}
  for name in ("palindrome", "failed_right_dp", "restart_search", "chain_shift",
               "nonchain_move", "end_chain", "end_pending"):
    state = p.add(("halt",))
    setattr(p, name, state)
    p.outcomes[state] = name
  p.no_chain = search.missed
  # These remain available to composition clients expecting search labels.
  p.found, p.missed = search.found, search.missed
  ready = {(phase, direction): p.reserve()
           for phase in range(5) for direction in (-1, 1)}
  p.ready = set(ready.values())
  p.confirmed_ready = {q for (phase, _), q in ready.items() if phase == 4}
  letters = {token: s for s in alphabet
             for token in (s, center_symbol(s), period_symbol(s))}
  period_letters = {token: s for s in alphabet for token in (s, first(s), last(s))}

  for (phase, direction), q in ready.items():
    choices = {BLANK: q, END: p.write(PORT, BLANK,
               p.end_chain if phase == 4 else p.end_pending)}
    for right in alphabet:
      compare_period = {}
      for token, prediction in period_letters.items():
        if token.startswith("F:"):
          next_direction, next_phase = 1, min(4, phase + 1)
        elif token.startswith("T:"):
          next_direction, next_phase = -1, min(4, phase + 1)
        else:
          next_direction, next_phase = direction, phase
        advance = p.move(PERIOD, next_direction, ready[next_phase, next_direction])
        boundary = p.branch(WINDOW, {LEFT: p.palindrome, **{
          symbol: advance for symbol in letters}})
        matching = p.move(WINDOW, -1, boundary)
        compare_left = {}
        for symbol, left in letters.items():
          if left == right == prediction:
            target = matching
          elif phase < 4:
            target = p.failed_right_dp
          elif left == right:
            target = p.restart_search
          elif prediction == right:
            target = p.chain_shift
          else:
            target = p.nonchain_move
          compare_left[symbol] = p.write(PORT, BLANK, target)
        compare_period[token] = p.branch(WINDOW, compare_left)
      choices[right] = p.branch(PERIOD, compare_period)
    p.branch(PORT, choices, q)

  # Read the first marked semiperiod directly from the DP result: C..C-h.
  # A local bouncing reader avoids O(h) rewinds between incoming places.
  back_period, back_window, copy_period = p.reserve(), p.reserve(), p.reserve()
  start_matching = p.move(WINDOW, -1, p.move(PERIOD, 1, ready[0, 1]))
  p.branch(WINDOW, {**{center_symbol(s): start_matching for s in alphabet}, **{
    token: p.move(WINDOW, 1, back_window)
    for s in alphabet for token in (s, period_symbol(s))}}, back_window)
  p.branch(PERIOD, {LEFT: p.move(PERIOD, 1, back_window), **{
    token: p.move(PERIOD, -1, back_period) for token in period_letters}}, back_period)
  p.branch(WINDOW, {**{
    period_symbol(s): p.write(PERIOD, last(s), back_period) for s in alphabet}, **{
    s: p.write(PERIOD, s, p.move(PERIOD, 1,
       p.move(WINDOW, -1, copy_period))) for s in alphabet}}, copy_period)
  initial_symbol = p.branch(WINDOW, {
    center_symbol(s): p.write(PERIOD, first(s), p.move(PERIOD, 1,
      p.move(WINDOW, -1, copy_period))) for s in alphabet})
  p.code[search.found] = ("write", PERIOD, LEFT, p.move(PERIOD, 1, initial_symbol))
  p.start = p.write(PORT, BLANK, search.start)
  p.validate()
  return p

File: scala/pal/src/main/scala/pal/ChainFinite.scala
package pal

import scala.collection.immutable.VectorMap

import FppFinite.*
import FppFinite.Instruction.*
import DpSearchFinite.{buildSearchProgram, WINDOW, STATUS, centerSymbol, periodSymbol}

/** Finite right-dp / three-way chain monitor, composed after local DP search.
  *
  * The caller supplies right-side places C+1,C+2,... through a one-cell PORT,
  * only at a ready state. Setup is offline. Each subsequent place takes
  * constant local work: PERIOD bounces along a marked semiperiod, while
  * WINDOW moves left for ordinary palindrome matching. This implements the
  * branch decisions, not the outer center shift/restart/move scheduler.
  *
  * Python 原典: `chain_finite.py`。
  */
object ChainFinite {
  val PERIOD: Int = 16
  val PORT: Int = 17

  /** The outcome halt states, in the order Python's `build_chain_program` allocates them. */
  val OUTCOME_NAMES: Vector[String] = Vector("palindrome", "failed_right_dp", "restart_search",
    "chain_shift", "nonchain_move", "end_chain", "end_pending")

  /** The first symbol of the copied semiperiod (PERIOD bounces right from here). */
  def first(symbol: String): String = "F:" + symbol

  /** The last symbol of the copied semiperiod (PERIOD bounces left from here). */
  def last(symbol: String): String = "T:" + symbol

  /** Python の `build_chain_program(alphabet="abs")`。
    *
    * @param bodySymbols `FppReuse.makeReusable` に渡す掃除記号の順序（ゴールデン再生成用）。
    */
  def buildChainProgram(alphabet: String = "abs", bodySymbols: Option[Seq[String]] = None): Program = {
    val search = buildSearchProgram(alphabet, bodySymbols)
    val p = new Program(search.alphabet, ntapes = 18)
    p.sourceAlphabet = search.sourceAlphabet
    p.scalarTapes = Some(Vector(STATUS, PORT))
    p.copyCodeFrom(search)
    val letters = p.sourceAlphabet
    p.outcomes(search.missed.get) = "no_chain"
    val outcome: Map[String, Int] = OUTCOME_NAMES.map { name =>
      val state = p.add(Halt)
      p.outcomes(state) = name
      name -> state
    }.toMap
    // These remain available to composition clients expecting search labels.
    p.found = search.found
    p.missed = search.missed
    val ready: VectorMap[(Int, Int), Int] = VectorMap.from(for {
      phase <- 0 until 5
      direction <- Seq(-1, 1)
    } yield (phase, direction) -> p.reserve())
    p.ready = ready.values.toSet
    p.confirmedReady = ready.collect { case ((phase, _), q) if phase == 4 => q }.toSet
    val tokenLetters: VectorMap[String, String] = VectorMap.from(for {
      s <- letters
      token <- Seq(s, centerSymbol(s), periodSymbol(s))
    } yield token -> s)
    val periodLetters: VectorMap[String, String] = VectorMap.from(for {
      s <- letters
      token <- Seq(s, first(s), last(s))
    } yield token -> s)

    for (((phase, direction), q) <- ready) {
      var portChoices = choices(BLANK -> q, END -> p.write(PORT, BLANK,
        if (phase == 4) { outcome("end_chain") } else { outcome("end_pending") }))
      for (right <- letters) {
        var comparePeriod = VectorMap.empty[String, Int]
        for ((token, prediction) <- periodLetters) {
          val (nextDirection, nextPhase) =
            if (token.startsWith("F:")) { (1, math.min(4, phase + 1)) }
            else if (token.startsWith("T:")) { (-1, math.min(4, phase + 1)) }
            else { (direction, phase) }
          val advance = p.move(PERIOD, nextDirection, ready((nextPhase, nextDirection)))
          val boundary = p.branch(WINDOW, choices((Seq(LEFT -> outcome("palindrome")) ++
            tokenLetters.keys.map(_ -> advance))*))
          val matching = p.move(WINDOW, -1, boundary)
          var compareLeft = VectorMap.empty[String, Int]
          for ((symbol, left) <- tokenLetters) {
            val target =
              if (left == right && right == prediction) { matching }
              else if (phase < 4) { outcome("failed_right_dp") }
              else if (left == right) { outcome("restart_search") }
              else if (prediction == right) { outcome("chain_shift") }
              else { outcome("nonchain_move") }
            compareLeft = compareLeft.updated(symbol, p.write(PORT, BLANK, target))
          }
          comparePeriod = comparePeriod.updated(token, p.branch(WINDOW, compareLeft))
        }
        portChoices = portChoices.updated(right, p.branch(PERIOD, comparePeriod))
      }
      p.branch(PORT, portChoices, q)
    }

    // Read the first marked semiperiod directly from the DP result: C..C-h.
    // A local bouncing reader avoids O(h) rewinds between incoming places.
    val backPeriod = p.reserve()
    val backWindow = p.reserve()
    val copyPeriod = p.reserve()
    val startMatching = p.move(WINDOW, -1, p.move(PERIOD, 1, ready((0, 1))))
    p.branch(WINDOW, choices((letters.map(s => centerSymbol(s) -> startMatching) ++
      letters.flatMap(s => Seq(s, periodSymbol(s)).map(token => token -> p.move(WINDOW, 1, backWindow))))*), backWindow)
    p.branch(PERIOD, choices((Seq(LEFT -> p.move(PERIOD, 1, backWindow)) ++
      periodLetters.keys.map(token => token -> p.move(PERIOD, -1, backPeriod)))*), backPeriod)
    p.branch(WINDOW, choices((letters.map(s => periodSymbol(s) -> p.write(PERIOD, last(s), backPeriod)) ++
      letters.map(s => s -> p.write(PERIOD, s, p.move(PERIOD, 1, p.move(WINDOW, -1, copyPeriod)))))*), copyPeriod)
    val initialSymbol = p.branch(WINDOW, choices(letters.map { s =>
      centerSymbol(s) -> p.write(PERIOD, first(s), p.move(PERIOD, 1, p.move(WINDOW, -1, copyPeriod)))
    }*))
    p.set(search.found.get, Write(PERIOD, LEFT, p.move(PERIOD, 1, initialSymbol)))
    p.start = p.write(PORT, BLANK, search.start)
    p.validate()
    p
  }
}
